infer/schema.go

```diff
@@ -127,6 +128,11 @@ func serializeTypeAsPropertyType(
 			Ref: "pulumi.json#/Archive",
 		}, nil
 	}
+	if t == reflect.TypeOf(types.AssetOrArchive{}) {
+		return schema.TypeSpec{
+			Ref: "pulumi.json#/Asset",
+		}, nil
+	}
```

Comment:
Should we add a warning here if they're using the previous Asset-only type above as it doesn't round-trip and they should migrate to AssetOrArchive?

Reply:
<img width="1532" alt="Screenshot 2024-06-11 at 20 39 46" src="https://github.com/pulumi/pulumi-go-provider/assets/12074/384f5c23-ab88-4b64-bf14-c3266f0ca1ec">

Reply:
Hmm, I just noticed that this also causes:
```
$ pulumi package get-schema /Users/tkappler/pulumi/pulumi-command/bin/pulumi-resource-command
warning: Please use pulumi-go-provider's types.AssetOrArchive type instead of resource.Asset or resource.Archive. ...
```

The reason is that types are serialized recursively into the schema, so first `AssetOrArchive`, then its inner `Asset` and `Archive` types.

Reply:
Let's leave a warning as a follow up PR.

Reply:
https://github.com/pulumi/pulumi-go-provider/issues/243